### PR TITLE
First step towards getting java8 workaround for maven.

### PR DESCRIPTION
**Pull request by danielmoy-google**
schroederc has #3093 in, which breaks maven/gradle stuffs, need to fix
forward.

### CODE REVIEW

**Comment by danielmoy-google** [comment]
Ah nope, this is insufficient.  It did fix the first bit:

```
$ docker run  -v /tmp/someoutdir/:/tmp/someoutdir/ -v /tmp/somecoderepo/:/repo \
  -it extractiondockerfile1234567 /bin/bash

# jar tf /opt/kythe/extractors/javac8_extractor.jar | grep com/sun/tools/javac | wc -l
1244
```

But more problems remain:

```
# TMPDIR=/tmp/waterr \
KYTHE_ROOT_DIRECTORY=/repo \
KYTHE_OUTPUT_DIRECTORY=/tmp/watout \
REAL_JAVAC=`which javac` \
JAVAC_EXTRACTOR_JAR=/opt/kythe/extractors/javac8_extractor.jar \
mvn clean install \
-Dmaven.compiler.forceJavacCompilerUse=true \
-Dmaven.compiler.fork=true \
-Dmaven.compiler.executable=/opt/kythe/extractors/javac-wrapper.sh

<bunch of output>

# ls /tmp/watout
(nothing)

# cat /tmp/waterr/javac-extractor.err 
no main manifest attribute, in /opt/kythe/extractors/javac8_extractor.jar
no main manifest attribute, in /opt/kythe/extractors/javac8_extractor.jar
```

File: kythe/java/com/google/devtools/kythe/extractors/java/artifacts/Dockerfile [hunk 1]
```diff
@@ -22,3 +22,5 @@ FROM debian:jessie
 ADD kythe/go/extractors/config/runextractor/runextractor /opt/kythe/extractors/runextractor
 ADD kythe/java/com/google/devtools/kythe/extractors/java/standalone/javac-wrapper.sh /opt/kythe/extractors/javac-wrapper.sh
 ADD kythe/java/com/google/devtools/kythe/extractors/java/standalone/javac_extractor_deploy.jar /opt/kythe/extractors/javac_extractor.jar
+# Yea ok the thing for running java8 says java9, oh well.
+ADD third_party/javac/javac-9-dev-r4023-1.jar /opt/kythe/extractors/javac8_extractor.jar
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/extractors/java/artifacts/Dockerfile]: The name is correct.  This is the javac langtools jar for Java 9.  It's compiled to target JRE 8.

**Comment by schroederc** [comment]
> # TMPDIR=/tmp/waterr \
> KYTHE_ROOT_DIRECTORY=/repo \
> KYTHE_OUTPUT_DIRECTORY=/tmp/watout \
> REAL_JAVAC=`which javac` \
> JAVAC_EXTRACTOR_JAR=/opt/kythe/extractors/javac8_extractor.jar \
> mvn clean install \
> -Dmaven.compiler.forceJavacCompilerUse=true \
> -Dmaven.compiler.fork=true \
> -Dmaven.compiler.executable=/opt/kythe/extractors/javac-wrapper.sh

The command listed above is incorrect.  `JAVAC_EXTRACTOR_JAR` should still be the Kythe extractor jar.  `KYTHE_JAVA_RUNTIME_OPTIONS` should include `-Xbootclasspath/p:javac-9-dev-r4023-1.jar`

File: kythe/go/extractors/config/base/mvn/mvn.go [hunk 2]
```diff
@@ -70,11 +70,15 @@ const DefaultConfig = `{
         },
         {
           "name": "JAVAC_EXTRACTOR_JAR",
-          "value": "/opt/kythe/extractors/javac_extractor.jar"
+          "value": "/opt/kythe/extractors/javac8_extractor.jar"
```
Comment (schroederc) [inline kythe/go/extractors/config/base/mvn/mvn.go]: This shouldn't change

File: kythe/go/extractors/config/base/testdata/expected_mvn_config.Dockerfile [hunk 3]
```diff
@@ -14,11 +14,12 @@ ENV JAVA_HOME=/docker-java-home
 ENV PATH=$JAVA_HOME/bin:$PATH
 COPY --from=javac-extractor-artifacts /opt/kythe/extractors/runextractor /opt/kythe/extractors/runextractor
 COPY --from=javac-extractor-artifacts /opt/kythe/extractors/javac-wrapper.sh /opt/kythe/extractors/javac-wrapper.sh
-COPY --from=javac-extractor-artifacts /opt/kythe/extractors/javac_extractor.jar /opt/kythe/extractors/javac_extractor.jar
+COPY --from=javac-extractor-artifacts /opt/kythe/extractors/javac8_extractor.jar /opt/kythe/extractors/javac8_extractor.jar
```
Comment (schroederc) [inline kythe/go/extractors/config/base/testdata/expected_mvn_config.Dockerfile]: This should be in addition to `javac_extractor`.

File: kythe/go/extractors/config/base/testdata/expected_mvn_config.Dockerfile [hunk 4]
```diff
@@ -14,11 +14,12 @@ ENV JAVA_HOME=/docker-java-home
...
 ENV KYTHE_CORPUS=testcorpus
 ENV REAL_JAVAC=$JAVA_HOME/bin/javac
-ENV JAVAC_EXTRACTOR_JAR=/opt/kythe/extractors/javac_extractor.jar
+ENV JAVAC_EXTRACTOR_JAR=/opt/kythe/extractors/javac8_extractor.jar
```
Comment (schroederc) [inline kythe/go/extractors/config/base/testdata/expected_mvn_config.Dockerfile]: This should not change

File: kythe/go/extractors/config/base/testdata/expected_mvn_config.Dockerfile [hunk 5]
```diff
@@ -14,11 +14,12 @@ ENV JAVA_HOME=/docker-java-home
...
 ENV KYTHE_OUTPUT_FILE=$KYTHE_OUTPUT_DIRECTORY/extractor-output.kzip
+ENV KYTHE_JAVA_RUNTIME_OPTIONS=-Xbootclasspath/p:$JAVAC_EXTRACTOR_JAR
```
Comment (schroederc) [inline kythe/go/extractors/config/base/testdata/expected_mvn_config.Dockerfile]: This should point to `javac*.jar`.

File: kythe/go/extractors/config/base/testdata/expected_mvn_config.Dockerfile (hunk 3 above)
Comment (danielmoy-google) [reply kythe/go/extractors/config/base/testdata/expected_mvn_config.Dockerfile]: done

File: kythe/go/extractors/config/base/testdata/expected_mvn_config.Dockerfile (hunk 4 above)
Comment (danielmoy-google) [reply kythe/go/extractors/config/base/testdata/expected_mvn_config.Dockerfile]: done

File: kythe/go/extractors/config/base/testdata/expected_mvn_config.Dockerfile (hunk 5 above)
Comment (danielmoy-google) [reply kythe/go/extractors/config/base/testdata/expected_mvn_config.Dockerfile]: done

File: kythe/java/com/google/devtools/kythe/extractors/java/artifacts/Dockerfile (hunk 1 above)
Comment (danielmoy-google) [reply kythe/java/com/google/devtools/kythe/extractors/java/artifacts/Dockerfile]: ok fair enough.  also your suggestion of java9_tools.jar is more clear

File: kythe/go/extractors/config/base/mvn/mvn.go (hunk 2 above)
Comment (danielmoy-google) [reply kythe/go/extractors/config/base/mvn/mvn.go]: done

**Comment by danielmoy-google** [comment]
current guess is that mvn needs to have a maven.compiler.source/target that is explicitly set

the fact that I didn't package any editor into the docker image was, in retrospect, a dumb decision.  no vi/vim/emacs/nano/ed/?? to edit files with.

**Comment by danielmoy-google** [comment]
> current guess is that mvn needs to have a maven.compiler.source/target that is explicitly set
> 
> the fact that I didn't package any editor into the docker image was, in retrospect, a dumb decision. no vi/vim/emacs/nano/ed/?? to edit files with.

Well that was indeed sufficient to make the error go away:

```
# rm -f /tmp/waterr/* /tmp/watout/* && TMPDIR=/tmp/waterr KYTHE_ROOT_DIRECTORY=/repo KYTHE_OUTPUT_DIRECTORY=/tmp/watout REAL_JAVAC=`which javac` JAVAC_EXTRACTOR_JAR=/opt/kythe/extractors/javac_extractor.jar KYTHE_JAVA_RUNTIME_OPTIONS=-Xbootclasspath/p:/opt/kythe/extractors/javac8_extractor.jar mvn clean install -Dmaven.compiler.source=1.6 -Dmaven.compiler.target=1.6 -Dmaven.compiler.forceJavacCompilerUse=true -Dmaven.compiler.fork=true -Dmaven.compiler.executable=/opt/kythe/extractors/javac-wrapper.sh

# cat /tmp/waterr/javac-extractor.err
(nothing)
```

but still no output, so that wasn't the right idea:

```
# ls /tmp/watout/
(nothing)
```

**Comment by danielmoy-google** [comment]
> > current guess is that mvn needs to have a maven.compiler.source/target that is explicitly set
> > the fact that I didn't package any editor into the docker image was, in retrospect, a dumb decision. no vi/vim/emacs/nano/ed/?? to edit files with.
> 
> Well that was indeed sufficient to make the error go away:
> 
> ```
> # rm -f /tmp/waterr/* /tmp/watout/* && TMPDIR=/tmp/waterr KYTHE_ROOT_DIRECTORY=/repo KYTHE_OUTPUT_DIRECTORY=/tmp/watout REAL_JAVAC=`which javac` JAVAC_EXTRACTOR_JAR=/opt/kythe/extractors/javac_extractor.jar KYTHE_JAVA_RUNTIME_OPTIONS=-Xbootclasspath/p:/opt/kythe/extractors/javac8_extractor.jar mvn clean install -Dmaven.compiler.source=1.6 -Dmaven.compiler.target=1.6 -Dmaven.compiler.forceJavacCompilerUse=true -Dmaven.compiler.fork=true -Dmaven.compiler.executable=/opt/kythe/extractors/javac-wrapper.sh
> 
> # cat /tmp/waterr/javac-extractor.err
> (nothing)
> ```
> but still no output, so that wasn't the right idea:
> 
> ```
> # ls /tmp/watout/
> (nothing)
> ```

Ah ah!

But actually we don't care if this works!  Because lo and behold, my binary runextractor already goes in and mucks with pom.xml to set the right things:

Sure enough:

```
# /opt/kythe/extractors/runextractor maven -pom_xml pom.xml -javac_wrapper /opt/kythe/extractors/javac-wrapper.sh

# ls -l /tmp/realoutdir/
total 12
-rw-r--r-- 1 root root 9873 Oct  2 23:44 extractor-output.kzip
```

yea :)

**Pull request merged**